src/java/com/rapleaf/jack/queries/GenericQueryBuilder.java

```diff
@@ -0,0 +1,165 @@
+    this.dbConnection = dbConnection;
+    this.genericQuery = genericQuery;
+  }
+
+  public GenericQueryBuilder leftJoin(Class<? extends ModelWithId> model, ModelField modelField1, ModelField modelField2) {
+    genericQuery.addJoinCondition(new JoinCondition(JoinType.LEFT_JOIN, model, modelField1, modelField2));
+    return this;
+  }
+
+  public GenericQueryBuilder rightJoin(Class<? extends ModelWithId> model, ModelField modelField1, ModelField modelField2) {
+    genericQuery.addJoinCondition(new JoinCondition(JoinType.RIGHT_JOIN, model, modelField1, modelField2));
+    return this;
+  }
+
+  public GenericQueryBuilder innerJoin(Class<? extends ModelWithId> model, ModelField modelField1, ModelField modelField2) {
+    genericQuery.addJoinCondition(new JoinCondition(JoinType.INNER_JOIN, model, modelField1, modelField2));
+    return this;
+  }
+
+  public GenericQueryBuilder where(ModelField modelField, IWhereOperator operator) {
+    genericQuery.addWhereCondition(new WhereConstraint(modelField, operator));
+    return this;
+  }
+
+  public GenericQueryBuilder orderBy(ModelField modelField, QueryOrder queryOrder) {
+    genericQuery.addOrderCondition(new OrderCriterion(modelField, queryOrder));
+    return this;
+  }
+
+  public GenericQueryBuilder orderBy(ModelField modelField) {
+    genericQuery.addOrderCondition(new OrderCriterion(modelField, QueryOrder.ASC));
+    return this;
+  }
+
+  public GenericQueryBuilder limit(int offset, int limit) {
+    genericQuery.addLimitCondition(new LimitCriterion(offset, limit));
+    return this;
+  }
+
+  public GenericQueryBuilder limit(int limit) {
+    genericQuery.addLimitCondition(new LimitCriterion(0, limit));
+    return this;
+  }
+
+  public GenericQueryBuilder select(ModelField... modelFields) {
+    for (ModelField modelField : modelFields) {
+      genericQuery.addSelectedModelField(modelField);
+    }
+    return this;
+  }
+
+  public String getSqlStatement() {
+    return genericQuery.getSqlStatement();
+  }
+
+  public PreparedStatement getPreparedStatement() throws IOException {
+    PreparedStatement preparedStatement = dbConnection.getPreparedStatement(getSqlStatement());
+    setStatementParameters(preparedStatement);
+    return preparedStatement;
+  }
+
+  public List<QueryEntry> fetch() throws IOException {
+    int retryCount = 0;
+    PreparedStatement preparedStatement;
+    List<QueryEntry> results = Lists.newArrayList();
+
+    while (true) {
+      preparedStatement = getPreparedStatement();
+      try {
+        queryExecution(results, preparedStatement);
+        return results;
+      } catch (SQLRecoverableException e) {
+        if (++retryCount > MAX_CONNECTION_RETRIES) {
+          throw new IOException(e);
+        }
+      } catch (SQLException e) {
+        throw new IOException(e);
+      }
+    }
+  }
+
+  private void queryExecution(List<QueryEntry> results, PreparedStatement preparedStatement) throws SQLException {
```

Comment:
Since these are all select queries and always return results why have it be void and populate the list passed in instead of returning the results List? I actually think this could cause issues with the retry logic if the exception happened after results was partially populated.